controllers/machinedeployment_rolling_test.go

```diff
@@ -0,0 +1,221 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"strconv"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/controllers/mdutil"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+func TestReconcileNewMachineSet(t *testing.T) {
+	testCases := []struct {
+		name                          string
+		machineDeployment             *clusterv1.MachineDeployment
+		newMachineSet                 *clusterv1.MachineSet
+		allMachineSets                []*clusterv1.MachineSet
+		expectedNewMachineSetReplicas int
+		error                         error
+	}{
+		{
+			name: "It fails when machineDeployment has no replicas",
+			machineDeployment: &clusterv1.MachineDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+					Name:      "bar",
+				},
+			},
+			newMachineSet: &clusterv1.MachineSet{
+				Spec: clusterv1.MachineSetSpec{
+					Replicas: pointer.Int32Ptr(2),
+				},
+			},
+			error: errors.Errorf("spec replicas for MachineDeployment foo/bar is nil, this is unexpected"),
+		},
+		{
+			name: "It fails when new machineSet has no replicas",
+			machineDeployment: &clusterv1.MachineDeployment{
+				Spec: clusterv1.MachineDeploymentSpec{
+					Replicas: pointer.Int32Ptr(2),
+				},
+			},
+			newMachineSet: &clusterv1.MachineSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+					Name:      "bar",
+				},
+			},
+			error: errors.Errorf("spec replicas for MachineSet foo/bar is nil, this is unexpected"),
+		},
+		{
+			name: "RollingUpdate strategy: Scale up: 0 -> 2",
+			machineDeployment: &clusterv1.MachineDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: "foo",
+					Name:      "bar",
+				},
+				Spec: clusterv1.MachineDeploymentSpec{
+					Strategy: &clusterv1.MachineDeploymentStrategy{
+						Type: clusterv1.RollingUpdateMachineDeploymentStrategyType,
+						RollingUpdate: &clusterv1.MachineRollingUpdateDeployment{
+							MaxUnavailable: intOrStrPtr(0),
+							MaxSurge:       intOrStrPtr(2),
+						},
+					},
+					Replicas: pointer.Int32Ptr(2),
+				},
+			},
+			newMachineSet: &clusterv1.MachineSet{
```

Comment:
I'm pretty unfamiliar with this code, so I might be missing something entirely.

To me it looks like in allMS is always `append(oldMSs, newMS)`
https://github.com/kubernetes-sigs/cluster-api/blob/e9494401157daf5994eb21134ebaa73c86c4822b/controllers/machinedeployment_rolling.go#L45-L48

I think we should test also cases where this is the case. As far as I can see the current tests all have disjunct newMS/allMS, is this intended?

Reply:
This is actually already being ensured for every testCase, see https://github.com/kubernetes-sigs/cluster-api/pull/4495/files#diff-a70da8885988c24ea4a35f7c1291f24a99f8c5ffdedf2f27864b90575afc2312R188

Reply:
I just renamed the field allMachineSets -> OldMachineSets to make this more obvious.

Reply:
Oh, shouldn't have missed that :/

Reply:
you raised a good point which resulted in better naming, thanks!